File: backend/app/schemas/subscription.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from decimal import Decimal
from app.models.subscription import PlanTier, SubscriptionStatus
from app.schemas.merchant import MerchantResponse
from app.schemas.invoice import InvoiceResponse

class SubscriptionPlanResponse(BaseModel):
    id: int
    tier: PlanTier
    name: str
    price: Decimal
    duration_days: int
    is_active: bool

    class Config:
        from_attributes = True

class SubscriptionRequest(BaseModel):
    plan_id: int
    auto_renew: bool = True

class SubscriptionResponse(BaseModel):
    id: int
    merchant_id: int
    plan_id: int
    status: SubscriptionStatus
    current_period_start: Optional[datetime]
    current_period_end: Optional[datetime]
    auto_renew: bool
    
    class Config:
        from_attributes = True

class SubscriptionCheckoutResponse(BaseModel):
    subscription: SubscriptionResponse
    verification_url: str
    token: str

class SubscriptionPlanCreate(BaseModel):
    tier: PlanTier
    name: str
    price: Decimal = Field(..., ge=0, decimal_places=2)
    duration_days: int = Field(..., gt=0)
    is_active: bool = True

class SubscriptionPlanUpdate(BaseModel):
    tier: Optional[PlanTier] = None
    name: Optional[str] = None
    price: Optional[Decimal] = Field(None, ge=0, decimal_places=2)
    duration_days: Optional[int] = Field(None, gt=0)

class SubscriptionWithDetailsResponse(BaseModel):
    id: int
    merchant_id: int
    plan_id: int
    status: SubscriptionStatus
    current_period_start: Optional[datetime]
    current_period_end: Optional[datetime]
    auto_renew: bool
    created_at: datetime
    updated_at: datetime
    
    merchant: Optional[MerchantResponse] = None
    plan: Optional[SubscriptionPlanResponse] = None
    invoices: Optional[List['SubscriptionInvoiceResponse']] = None
    
    class Config:
        from_attributes = True

class SubscriptionInvoiceResponse(BaseModel):
    id: int
    subscription_id: int
    invoice_id: int
    is_processed: bool
    created_at: datetime
    
    subscription: Optional[SubscriptionWithDetailsResponse] = None
    invoice: Optional[InvoiceResponse] = None
    
    class Config:
        from_attributes = True

class PlanRevenueResponse(BaseModel):
    plan_name: str
    tier: PlanTier
    total_revenue: Decimal

class SubscriptionAnalyticsResponse(BaseModel):
    total_active: int
    total_pending: int
    total_expired: int
    total_cancelled: int
    new_this_period: int
    total_revenue: Decimal
    conversion_rate: float
    revenue_by_plan: List[PlanRevenueResponse]
    period_start: datetime
    period_end: datetime

# Forward references
SubscriptionWithDetailsResponse.model_rebuild()
SubscriptionInvoiceResponse.model_rebuild()

File: admin-dashboard/src/types/subscription.ts
export type PlanTier = 'free' | 'starter' | 'professional' | 'enterprise';

export interface SubscriptionPlan {
  id: number;
  tier: PlanTier;
  name: string;
  price: number;
  duration_days: number;
  is_active: boolean;
  created_at: string;
}

export interface SubscriptionPlanCreate {
  tier: PlanTier;
  name: string;
  price: number;
  duration_days: number;
  is_active?: boolean;
}

export interface SubscriptionPlanUpdate {
  tier?: PlanTier;
  name?: string;
  price?: number;
  duration_days?: number;
}

export interface Subscription {
  id: number;
  merchant_id: number;
  plan_id: number;
  status: 'pending' | 'active' | 'expired' | 'cancelled';
  current_period_start: string | null;
  current_period_end: string | null;
  auto_renew: boolean;
  created_at: string;
  updated_at: string;
  merchant?: Merchant;
  plan?: SubscriptionPlan;
  invoices?: SubscriptionInvoice[];
}

export interface SubscriptionInvoice {
  id: number;
  subscription_id: number;
  invoice_id: number;
  is_processed: boolean;
  created_at: string;
  subscription?: Subscription;
  invoice?: Invoice;
}

export interface Merchant {
  id: number;
  user_id: number;
  is_active: boolean;
  profile: MerchantProfile;
  created_at: string;
}

export interface MerchantProfile {
  id: number;
  business_name: string;
  business_email: string;
  phone_number?: string;
}

export interface PlanRevenue {
  plan_name: string;
  tier: PlanTier;
  total_revenue: number;
}

export interface SubscriptionAnalytics {
  total_active: number;
  total_pending: number;
  total_expired: number;
  total_cancelled: number;
  new_this_period: number;
  total_revenue: number;
  conversion_rate: number;
  revenue_by_plan: PlanRevenue[];
  period_start: string;
  period_end: string;
}

export interface PaginatedSubscriptions {
  items: Subscription[];
  total: number;
}